### PR TITLE
Upload semantic-segmentation-deeplabv3

### PR DESCRIPTION
**Pull request by zyh911**
Add semantic-segmentation-deeplabv3 folder

### CODE REVIEW

File: semantic-segmentation-deeplabv3/module_specs/score.yaml [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        - opencv-python==4.1.0.25
+        - https://download.pytorch.org/whl/cu100/torch-1.1.0-cp36-cp36m-linux_x86_64.whl
+        - https://download.pytorch.org/whl/cu100/torchvision-0.3.0-cp36-cp36m-linux_x86_64.whl
+        - pillow==6.0.0
+        - pyarrow==0.14.0
+        - git+https://github.com/zyh911/CustomModules.git@t-yuhazh/add_semantic_segmentation_deeplabv3#subdirectory=semantic-segmentation-deeplabv3
+    command:
+    - python
+    - -m
+    - script.score
```
Comment (chjinche) [inline semantic-segmentation-deeplabv3/module_specs/score.yaml]: Would be better to change package name 'script' because it looks improper for a package

File: semantic-segmentation-deeplabv3/script/dataset.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
+import os
+import random
+import numpy as np
+from PIL import Image, ImageOps, ImageFilter
+
```
Comment (chjinche) [inline semantic-segmentation-deeplabv3/script/dataset.py]: Remove blank line

File: semantic-segmentation-deeplabv3/script/dataset.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+import torch
+
+
+class SegmentationDataset(object):
+    """Segmentation Base Dataset"""
```
Comment (chjinche) [inline semantic-segmentation-deeplabv3/script/dataset.py]: Refer to docstring to modify comment format

File: semantic-segmentation-deeplabv3/script/dataset.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        return self.NUM_CLASS
+
+    @property
+    def pred_offset(self):
+        return 0
+
+
+class VOCSegmentation(SegmentationDataset):
+    """Pascal VOC Semantic Segmentation Dataset.
+    Parameters
```
Comment (chjinche) [inline semantic-segmentation-deeplabv3/script/dataset.py]: Add a blank line

File: semantic-segmentation-deeplabv3/script/metric.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+import torch
+import numpy as np
+
+__all__ = ['SegmentationMetric', 'batch_pix_accuracy', 'batch_intersection_union',
+           'pixelAccuracy', 'intersectionAndUnion', 'hist_info', 'compute_score']
+
+
+class SegmentationMetric(object):
+    """
+    Computes pixAcc and mIoU metric scores
```
Comment (chjinche) [inline semantic-segmentation-deeplabv3/script/metric.py]: Docstring format

File: semantic-segmentation-deeplabv3/script/metric.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+    def __init__(self, nclass):
+        super(SegmentationMetric, self).__init__()
+        self.nclass = nclass
+        self.reset()
+
+    def evaluate_worker(self, pred, label):
+        correct, labeled = batch_pix_accuracy(pred, label)
+        inter, union = batch_intersection_union(pred, label, self.nclass)
+
```
Comment (chjinche) [inline semantic-segmentation-deeplabv3/script/metric.py]: Remove blank line

File: semantic-segmentation-deeplabv3/script/metric.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+    def update(self, preds, labels):
+        """Updates the internal evaluation result.
+        Parameters
+        ----------
+        labels : 'NumpyArray' or list of `NumpyArray`
+            The labels of the data.
+        preds : 'NumpyArray' or list of `NumpyArray`
+            Predicted values.
+        """
+
```
Comment (chjinche) [inline semantic-segmentation-deeplabv3/script/metric.py]: Remove blank line

File: semantic-segmentation-deeplabv3/script/metric.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+        self.total_label += labeled
+        if self.total_inter.device != inter.device:
+            self.total_inter = self.total_inter.to(inter.device)
+            self.total_union = self.total_union.to(union.device)
+        self.total_inter += inter
+        self.total_union += union
+
+    def update(self, preds, labels):
+        """Updates the internal evaluation result.
+        Parameters
```
Comment (chjinche) [inline semantic-segmentation-deeplabv3/script/metric.py]: Add blank line

File: semantic-segmentation-deeplabv3/script/metric.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+        self.total_label = 0
+
+
+# pytorch version
+def batch_pix_accuracy(output, target):
+    """PixAcc"""
+    # inputs are numpy array, output 4D, target 3D
+    predict = torch.argmax(output.long(), 1) + 1
+    target = target.long() + 1
+
```
Comment (chjinche) [inline semantic-segmentation-deeplabv3/script/metric.py]: Remove

File: semantic-segmentation-deeplabv3/script/model.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
+import sys
+
+import torch.nn as nn
+from torchvision.models.segmentation.segmentation import _segm_resnet, model_urls
+try:
+    from torch.hub import load_state_dict_from_url
+except ImportError:
+    from torch.utils.model_zoo import load_url as load_state_dict_from_url
```
Comment (chjinche) [inline semantic-segmentation-deeplabv3/script/model.py]: How about just use "from torch.utils.model_zoo import load_url as load_state_dict_from_url"?

File: semantic-segmentation-deeplabv3/script/model.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+        else:
+            state_dict = load_state_dict_from_url(model_url, model_dir=model_path)
+            model.load_state_dict(state_dict)
+    return model
+
+
+class SegmentationFuncs:
+
+    @staticmethod
+    def fcn_resnet50(model_path='script/saved_model', pretrained=False,
```
Comment (chjinche) [inline semantic-segmentation-deeplabv3/script/model.py]: Refer to https://github.com/hirofumi-s-friends/CustomModules/blob/chjinche/add-image-classification-densenet/densenet-img-cls/densenetimgcls/densenet.py to modify the way you add model_path

File: semantic-segmentation-deeplabv3/script/score.py [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+from io import BytesIO
+from argparse import ArgumentParser
+import base64
+import pandas as pd
+import pyarrow.parquet as pq  # noqa: F401
+from azureml.studio.modulehost.handler.port_io_handler import OutputHandler
+from azureml.studio.common.datatypes import DataTypes
+from azureml.studio.common.datatable.data_table import DataTable
+from azureml.studio.common.io.data_frame_utils import load_data_frame_from_directory
+
```
Comment (chjinche) [inline semantic-segmentation-deeplabv3/script/score.py]: Remove

File: semantic-segmentation-deeplabv3/script/score.py [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+import torch
+from torchvision import transforms
+
```
Comment (chjinche) [inline semantic-segmentation-deeplabv3/script/score.py]: Remove

File: semantic-segmentation-deeplabv3/script/score.py [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+from .model import SegmentationNet
+from .smt_fake import smt_fake_file
+
+
+class Score:
+    def __init__(self, model_path, meta={}):
+        self.pallete = [
+            [255, 255, 255],
```
Comment (chjinche) [inline semantic-segmentation-deeplabv3/script/score.py]: Could you please explain these numbs?

File: semantic-segmentation-deeplabv3/script/score.py [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+            [0, 60, 100],
+            [0, 80, 100],
+            [0, 0, 230],
+            [119, 11, 32],
+            [0, 0, 0]
+        ]
+
+        self.inference_transforms = transforms.Compose([
+            transforms.ToTensor(),
+            transforms.Normalize([.485, .456, .406], [.229, .224, .225]),
```
Comment (chjinche) [inline semantic-segmentation-deeplabv3/script/score.py]: Also these

File: semantic-segmentation-deeplabv3/script/score.py [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+        self.model = SegmentationNet(model_type=meta['Model type'], model_path=model_path, pretrained=False)
+        self.model.load_state_dict(torch.load(os.path.join(model_path, 'model.pth'), map_location='cpu'))
+        if meta['Use CUDA'] and torch.cuda.is_available():
+            self.model = self.model.cuda()
+
+        self.model.eval()
+
+    def run(self, input, meta=None):
+        my_list = []
+        for i in range(input.shape[0]):
```
Comment (chjinche) [inline semantic-segmentation-deeplabv3/script/score.py]: Use func like 'decode_image_str' to convert https://github.com/hirofumi-s-friends/CustomModules/blob/chjinche/add-image-classification-densenet/densenet-img-cls/densenetimgcls/score.py

File: semantic-segmentation-deeplabv3/script/train.py [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+from .metric import SegmentationMetric
+from .model import SegmentationNet
+from .smt_fake import smt_fake_model
+
+
+def val(model, val_loader, use_cuda, metric):
+    model.eval()
+    metric.reset()
+    total_iteration = len(val_loader)
+
```
Comment (chjinche) [inline semantic-segmentation-deeplabv3/script/train.py]: Remove useless blank lines

File: semantic-segmentation-deeplabv3/script/train.py [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+        if use_cuda and torch.cuda.is_available():
+            input = input.cuda()
+            target = target.cuda()
+
+        with torch.no_grad():
+            output = model(input)
+        metric.update(output['out'], target)
+        pixAcc, mIoU = metric.get()
+
+        print('Validation || current iteration / total iterations : [{} / {}], time : {:.2f}, '
```
Comment (chjinche) [inline semantic-segmentation-deeplabv3/script/train.py]: Use logger not print for log

File: semantic-segmentation-deeplabv3/setup.py [hunk 19]
```diff
@@ -0,0 +1,12 @@
+from setuptools import setup
+
+
+setup(
+    name='semantic-segmentation-deeplabv3',
+    version='0.0.1',
+    description='Deeplabv3 Model for Semantic Segmentation Task',
+    packages=['script'],
```
Comment (chjinche) [inline semantic-segmentation-deeplabv3/setup.py]: Use 'find_packages' like https://github.com/hirofumi-s-friends/CustomModules/blob/chjinche/add-image-classification-densenet/densenet-img-cls/setup.py

File: semantic-segmentation-deeplabv3/script/dataset.py [hunk 20]
```diff
@@ -0,0 +1,196 @@
...
+
+    def _mask_transform(self, mask):
+        target = np.array(mask).astype('int32')
+        target[target == 255] = -1
+        return torch.from_numpy(target).long()
+
+    @property
+    def classes(self):
+        """Category names."""
+        return ('background', 'airplane', 'bicycle', 'bird', 'boat', 'bottle',
```
Comment (chjinche) [inline semantic-segmentation-deeplabv3/script/dataset.py]: How about passing 'classes' as parameter to be more generic?

File: semantic-segmentation-deeplabv3/script/dataset.py [hunk 21]
```diff
@@ -0,0 +1,196 @@
...
+    def __init__(self, root, split, mode, transform, base_size=520, crop_size=480):
+        super(SegmentationDataset, self).__init__()
+        self.root = root
+        self.transform = transform
+        self.split = split
+        self.mode = mode if mode is not None else split
+        self.base_size = base_size
+        self.crop_size = crop_size
+
+    def _val_sync_transform(self, img, mask):
```
Comment (chjinche) [inline semantic-segmentation-deeplabv3/script/dataset.py]: How about moving funcs like this transform to a separate py?

File: semantic-segmentation-deeplabv3/script/model.py [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+
+
+__all__ = ['SegmentationNet']
+
+
+def load_model(arch_type, backbone, model_path, pretrained, num_classes, **kwargs):
+    aux_loss = True
+    temp_arch_type = arch_type
+    if temp_arch_type == 'deeplabv3':
+        temp_arch_type = temp_arch_type[:-2]
```
Comment (thy09) [inline semantic-segmentation-deeplabv3/script/model.py]: Directly set `temp_arch_type='deeplab'`?

File: semantic-segmentation-deeplabv3/script/model.py [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+    model = _segm_resnet(temp_arch_type, backbone, num_classes, aux_loss, **kwargs)
+    if pretrained:
+        arch = arch_type + '_' + backbone + '_coco'
+        model_url = model_urls[arch]
```
Comment (thy09) [inline semantic-segmentation-deeplabv3/script/model.py]: Perhaps a KeyError will be raised if arch is not valid key?

File: semantic-segmentation-deeplabv3/script/score.py [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+                output_predictions = output.argmax(0).cpu().numpy()
+                resultImg = np.zeros((output_predictions.shape[0], output_predictions.shape[1], 3), dtype=np.uint8)
+                for idx in range(len(self.pallete)):
+                    resultImg[output_predictions == idx] = self.pallete[idx]
+
+                resultImg1 = Image.fromarray(resultImg).resize(img.size)
+                # resultImg1.save(os.path.join('script/outputs2', '{}_mask.png'.format(i)))
+                resultImg2 = Image.blend(img, resultImg1, 0.5)
+                # resultImg2.save(os.path.join('script/outputs2', '{}_fusion.png'.format(i)))
+                imgByteArr1 = BytesIO()
```
Comment (thy09) [inline semantic-segmentation-deeplabv3/script/score.py]: Duplicated codes, perhaps you could use a function to simplify.

File: semantic-segmentation-deeplabv3/setup.py (hunk 19 above)
Comment (zyh911) [reply semantic-segmentation-deeplabv3/setup.py]: Done!

File: semantic-segmentation-deeplabv3/script/score.py (hunk 24 above)
Comment (zyh911) [reply semantic-segmentation-deeplabv3/script/score.py]: Done!

File: semantic-segmentation-deeplabv3/script/model.py (hunk 22 above)
Comment (zyh911) [reply semantic-segmentation-deeplabv3/script/model.py]: Done!

File: semantic-segmentation-deeplabv3/script/model.py (hunk 23 above)
Comment (zyh911) [reply semantic-segmentation-deeplabv3/script/model.py]: Done!

File: semantic-segmentation-deeplabv3/script/model.py (hunk 11 above)
Comment (zyh911) [reply semantic-segmentation-deeplabv3/script/model.py]: They are different. I cannot do it that way.

File: semantic-segmentation-deeplabv3/script/train.py (hunk 17 above)
Comment (zyh911) [reply semantic-segmentation-deeplabv3/script/train.py]: Done

File: semantic-segmentation-deeplabv3/script/score.py (hunk 15 above)
Comment (zyh911) [reply semantic-segmentation-deeplabv3/script/score.py]: Done

File: semantic-segmentation-deeplabv3/script/score.py (hunk 13 above)
Comment (zyh911) [reply semantic-segmentation-deeplabv3/script/score.py]: Done

File: semantic-segmentation-deeplabv3/script/score.py (hunk 12 above)
Comment (zyh911) [reply semantic-segmentation-deeplabv3/script/score.py]: Done

File: semantic-segmentation-deeplabv3/script/metric.py (hunk 9 above)
Comment (zyh911) [reply semantic-segmentation-deeplabv3/script/metric.py]: Done

File: semantic-segmentation-deeplabv3/script/metric.py (hunk 8 above)
Comment (zyh911) [reply semantic-segmentation-deeplabv3/script/metric.py]: Done

File: semantic-segmentation-deeplabv3/script/metric.py (hunk 7 above)
Comment (zyh911) [reply semantic-segmentation-deeplabv3/script/metric.py]: Done

File: semantic-segmentation-deeplabv3/script/metric.py (hunk 6 above)
Comment (zyh911) [reply semantic-segmentation-deeplabv3/script/metric.py]: Done

File: semantic-segmentation-deeplabv3/script/dataset.py (hunk 4 above)
Comment (zyh911) [reply semantic-segmentation-deeplabv3/script/dataset.py]: Done

File: semantic-segmentation-deeplabv3/script/dataset.py (hunk 2 above)
Comment (zyh911) [reply semantic-segmentation-deeplabv3/script/dataset.py]: Done

File: semantic-segmentation-deeplabv3/script/metric.py (hunk 5 above)
Comment (zyh911) [reply semantic-segmentation-deeplabv3/script/metric.py]: Done

File: semantic-segmentation-deeplabv3/script/dataset.py (hunk 3 above)
Comment (zyh911) [reply semantic-segmentation-deeplabv3/script/dataset.py]: Done

File: semantic-segmentation-deeplabv3/module_specs/score.yaml (hunk 1 above)
Comment (zyh911) [reply semantic-segmentation-deeplabv3/module_specs/score.yaml]: Done

File: semantic-segmentation-deeplabv3/script/model.py (hunk 10 above)
Comment (zyh911) [reply semantic-segmentation-deeplabv3/script/model.py]: Done

File: semantic-segmentation-deeplabv3/script/score.py (hunk 14 above)
Comment (zyh911) [reply semantic-segmentation-deeplabv3/script/score.py]: Done

File: semantic-segmentation-deeplabv3/script/train.py (hunk 18 above)
Comment (zyh911) [reply semantic-segmentation-deeplabv3/script/train.py]: Done!

File: semantic-segmentation-deeplabv3/script/score.py (hunk 16 above)
Comment (zyh911) [reply semantic-segmentation-deeplabv3/script/score.py]: Done!

File: semantic-segmentation-deeplabv3/script/dataset.py (hunk 21 above)
Comment (zyh911) [reply semantic-segmentation-deeplabv3/script/dataset.py]: Transform is part of dataset. I don't think it's necessary to separate it.

File: semantic-segmentation-deeplabv3/script/dataset.py (hunk 20 above)
Comment (zyh911) [reply semantic-segmentation-deeplabv3/script/dataset.py]: The class is dataset specifically for PASCAL VOC 2012. So the class labels are static. I don't think it's necessary to do this by passing patameters.
